akka-docs/rst/scala/code/docs/persistence/query/PersistenceQueryDocSpec.scala

```diff
@@ -0,0 +1,212 @@
+/*
+ * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+
+package docs.persistence.query
+
+import akka.actor._
+import akka.persistence.query.scaladsl.ReadJournal
+import akka.persistence.{ Recovery, PersistentActor }
+import akka.persistence.query._
+import akka.stream.{ FlowShape, ActorMaterializer }
+import akka.stream.scaladsl.FlowGraph
+import akka.stream.scaladsl.{ Flow, Sink, Source }
+import akka.testkit.AkkaSpec
+import akka.util.Timeout
+import docs.persistence.query.PersistenceQueryDocSpec.TheOneWhoWritesToQueryJournal
+
+import scala.collection.immutable
+import scala.concurrent.Future
+import scala.concurrent.duration.FiniteDuration
+import scala.concurrent.duration._
+
+object PersistenceQueryDocSpec {
+
+  implicit val timeout = Timeout(3.seconds)
+
+  class NoopReadJournal(sys: ExtendedActorSystem) extends ReadJournal {
+    override def query[T, M](q: Query[T, M], hints: Hint*): Source[T, M] =
+      Source.empty.mapMaterializedValue(_ => null.asInstanceOf[M])
+  }
+
+  //#my-read-journal
+  class MyReadJournal(system: ExtendedActorSystem) extends ReadJournal {
+
+    // TODO from config
+    private val defaulRefreshInterval: FiniteDuration = 3.seconds
+
+    override def query[T, M](q: Query[T, M], hints: Hint*): Source[T, M] =
+      q match {
+        case EventsByTag(tag, offset) ⇒
+          val props = MyEventsByTagPublisher.props(tag, offset, refreshInterval(hints))
+          Source.actorPublisher[EventEnvelope](props)
+            .mapMaterializedValue(_ ⇒ noMaterializedValue)
```

Comment:
Probably not the ideal spot to discuss, but the context of the discussion between @ktoso and @rkuhn is helpful.

I started to implement a `ReadJournal` today for my mongo journal and the `query` method seemed a bit off.  The casting wasn't great, but also I didn't really like communicating with the user that a query was not supported via exception.  More type-safety would be nice.

Intuitively this seems like a good spot for typeclasses.  I'm thinking that the (read) journal implementers would provide implicit objects as witnesses for the queries they support, e.g.
(Excuse the hastily put together pseudo-pseudocode, there must be a better signature to remove the extra T & M parameters on the trait)

```scala
package akka.persistence.query
abstract class ReadJournal { 
...
trait SupportedQueryWitness[A <: Query[T,M],T,M]
}

package my.read.journal
object SupportedQueries {
  implicit object allPids extends SupportedQueryWitness[AllPersistenceIds,String,Unit]
}
```

Then `ReadJournal.query` could require a witness for that `Query[T,M]` to be in implicit scope.  The downside of this I see is that the user of the read journal needs a compile-time dependency on the journal implementation, rather than the typical run-time approach.  Of course they may need this anyways for more interesting `Query[T,M]`s.

Anyways, just some initial reactions.  Thanks for all the work you do, it's greatly appreciated!